CleverTapSDK/Encryption/CTCryptMigrator.m

```diff
@@ -0,0 +1,319 @@
+#import <UIKit/UIKit.h>
+#import "CTLocalDataStore.h"
+#import "CTConstants.h"
+#import "CTPreferences.h"
+#import "CTLoginInfoProvider.h"
+#import "CTEncryptionManager.h"
+#import "CTPreferences.h"
+#import "CTUtils.h"
+#import "CTCryptMigrator.h"
+
+@interface CTCryptMigrator()
+
+@property (nonatomic, strong) CleverTapInstanceConfig *config;
+@property (nonatomic, strong) CTDeviceInfo *deviceInfo;
+@property (nonatomic, strong) NSArray *piiKeys;
+@property (nonatomic, strong) CTEncryptionManager *cryptManager;
+@property (nonatomic, assign) CleverTapEncryptionLevel lastEncryptionLevel;
+
+@end
+
+@implementation CTCryptMigrator
+
+- (instancetype)initWithConfig:(CleverTapInstanceConfig *)config
+                 andDeviceInfo:(CTDeviceInfo*)deviceInfo {
+    if (self = [super init]) {
+        _config = config;
+        _deviceInfo = deviceInfo;
+        _piiKeys = CLTAP_ENCRYPTION_PII_DATA;
+        _cryptManager = [[CTEncryptionManager alloc] initWithAccountID:_config.accountId encryptionLevel:_config.encryptionLevel];
+        if ([self isMigrationNeeded]) {
+            [self performMigration];
+        }
+    }
+    return self;
+}
+
+- (BOOL)isMigrationNeeded {
+    self.lastEncryptionLevel = (CleverTapEncryptionLevel)[CTPreferences getIntForKey:[CTUtils getKeyWithSuffix:kENCRYPTION_KEY accountID:_config.accountId]
+                                                                      withResetValue:CleverTapEncryptionNone];
+    
+    long migrationDone = [CTPreferences getIntForKey:[CTUtils getKeyWithSuffix:CLTAP_ENCRYPTION_MIGRATION_STATUS accountID:_config.accountId]
+                                       withResetValue:0];
+    
+    long encryptionDone = [CTPreferences getIntForKey:[CTUtils getKeyWithSuffix:CLTAP_ENCRYPTION_ALGORITHM accountID:_config.accountId]
+                                        withResetValue:0];
+    
+    // If encryption is done, mark migration as complete and return
+    if (encryptionDone) {
+        [CTPreferences putInt:1 forKey:[CTUtils getKeyWithSuffix:CLTAP_ENCRYPTION_MIGRATION_STATUS accountID:_config.accountId]];
```

Comment:
The migration flag is updated every time even if encryption and migration are done.